packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -188,25 +189,29 @@ export class ConnectionStateHandler {
         // Report telemetry after we set client id, but before transitioning to Connected state below!
         this.handler.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
 
-        const quorumClients = this.handler.quorumClients();
         // Check if we already processed our own join op through delta storage!
         // we are fetching ops from storage in parallel to connecting to ordering service
         // Given async processes, it's possible that we have already processed our own join message before
         // connection was fully established.
         // Note that we might be still initializing quorum - connection is established proactively on load!
-        if ((quorumClients !== undefined && quorumClients.getMember(details.clientId) !== undefined)
-            || connectionMode === "read"
-        ) {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        const alreadyProcessedOwnJoinOp = this.handler.quorumClients()?.getMember(details.clientId) !== undefined;
+        const writeConnection = connectionMode === "write";
+        if (writeConnection && !alreadyProcessedOwnJoinOp) {
             this.startJoinOpTimer();
+        } else {
+            // There should be no timer for 'read' connections,
+            // and if we processed our Join op we would have processed our previous Leave op prior to that
+            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "Unexpected timer state" */);
+
+            // For write connections, we are caught up already (since we've seen our Join op) and this will be immediate
```

Comment:
Based on the if condition above, this may not be true right. We can be in this else block if the connection is write and we haven't processed our join op.

Reply:
Ah okay, my bad. In the scenario I mentioned, we will go in the if condition. We will be here if it's a write connection and we have processed our join op.

Reply:
But this still may not be immediate. There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

Reply:
Also, please use `processed` our join op vs `seen` which mean different things.

Reply:
> There could have been more ops after the join op and in that case, we will wait for those ops to be processed before we move to connected state.

Yeah re-reading this change I came to the same conclusion, glad you saw it too, I was trying to figure out if I was missing something now or before when I wrote the code :P